File: leetcode/127.cpp
class Solution {
public:
	int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
		int ret = 0;
		map<string, vector<string>> graph;
		set<string> visited;
		wordList.push_back(beginWord);
		construct_graph(wordList, graph);
		ret = bfs(beginWord, endWord, graph, visited);
		return ret;
	}

	bool connect(string &word1, string &word2) {
		int count = 0;
		for (int i = 0; i < word1.size(); i++) {
			if (word1[i] != word2[i]) count++;
		}
		return count == 1;
	}

	void construct_graph(vector<string>& wordList, map<string, vector<string>>& graph) {
		for (int i = 0; i < wordList.size() - 1; i++) {
			for (int j = i + 1; j < wordList.size(); j++) {
				if (connect(wordList[i], wordList[j])) {
					graph[wordList[i]].push_back(wordList[j]);
					graph[wordList[j]].push_back(wordList[i]);
				}
			}
		}

	}

	int bfs(string &beginWord, string &endWord, map<string, vector<string>>& graph, set<string> &visited) {
		int ret = 0;
		queue<pair<string, int>> q;
		q.push(pair(beginWord, 1));
		visited.insert(beginWord);
		pair<string, int> cur;
		while (!q.empty()) {
			cur = q.front();
			q.pop();
			if (cur.first == endWord) return cur.second;
			for (int i = 0; i < graph[cur.first].size(); i++) {
				if (visited.find(graph[cur.first][i]) == visited.end()) {
					q.push(pair(graph[cur.first][i], cur.second + 1));
					visited.insert(graph[cur.first][i]);
				}
			}
		}
		return 0;
	}
};

File: leetcode/127.java

class Solution {
    class Pair {
        String first;
        int second;

        Pair(String first, int second) {
            this.first = first;
            this.second = second;
        }
    }

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        int ret = 0;
        Map<String, List<String>> graph = new HashMap<>();

        Set<String> visited = new HashSet<>();
        wordList.add(beginWord);

        for (int i = 0; i < wordList.size(); i++) {
            graph.put(wordList.get(i), new ArrayList<String>());
        }

        construct_graph(wordList, graph);

        return bfs(beginWord, endWord, graph, visited);
    }

    boolean connect(String word1, String word2) {
        int count = 0;
        for (int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) count++;
        }
        return count == 1;
    }

    void construct_graph(List<String> wordList, Map<String, List<String>> graph) {
        String istr;
        String jstr;
        for (int i = 0; i < wordList.size() - 1; i++) {
            for (int j = i + 1; j < wordList.size(); j++) {
                istr = wordList.get(i);
                jstr = wordList.get(j);
                if (connect(istr, jstr)) {
                    graph.get(istr).add(new String(jstr));
                    graph.get(jstr).add(new String(istr));
                }
            }
        }
    }

    int bfs(String beginWord, String endWord, Map<String, List<String>> graph, Set<String> visited) {
        int ret = 0;
        List<String> list;
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(beginWord, 1));
        visited.add(beginWord);
        Pair cur;
        while (0 != q.size()) {
            cur = q.remove();
            if (cur.first.equals(endWord)) return cur.second;
            list = graph.get(cur.first);
            for (int i = 0; i < list.size(); i++) {
                if (visited.contains(list.get(i)) == false) {
                    q.add(new Pair(list.get(i), cur.second + 1));
                    visited.add(list.get(i));
                }
            }
        }
        return 0;
    }
};